src/git-unix/ogit-http-fetch/main.ml

```diff
@@ -0,0 +1,169 @@
+module Store = Git_unix.Store
+module Sync = Git_unix.Sync (Store) (Git_curl)
+
+let () = Random.self_init ()
+let src = Logs.Src.create "ogit-http-fetch" ~doc:"logs binary event"
+
+module Log = (val Logs.src_log src : Logs.LOG)
+
+let pad n x =
+  if String.length x > n then x else x ^ String.make (n - String.length x) ' '
+
+let pp_header ppf (level, header) =
+  let level_style =
+    match level with
+    | Logs.App -> Logs_fmt.app_style
+    | Logs.Debug -> Logs_fmt.debug_style
+    | Logs.Warning -> Logs_fmt.warn_style
+    | Logs.Error -> Logs_fmt.err_style
+    | Logs.Info -> Logs_fmt.info_style
+  in
+  let level = Logs.level_to_string (Some level) in
+  Fmt.pf ppf "[%a][%a]"
+    (Fmt.styled level_style Fmt.string)
+    level (Fmt.option Fmt.string)
+    (Option.map (pad 10) header)
+
+let reporter ppf =
+  let report src level ~over k msgf =
+    let k _ =
+      over ();
+      k ()
+    in
+    let with_src_and_stamp h _ k fmt =
+      let dt = Mtime.Span.to_us (Mtime_clock.elapsed ()) in
+      Fmt.kpf k ppf
+        ("%s %a %a: @[" ^^ fmt ^^ "@]@.")
+        (pad 10 (Fmt.str "%+04.0fus" dt))
+        pp_header (level, h)
+        Fmt.(styled `Magenta string)
+        (pad 10 @@ Logs.Src.name src)
+    in
+    msgf @@ fun ?header ?tags fmt -> with_src_and_stamp header tags k fmt
+  in
+  { Logs.report }
+
+let setup_logs style_renderer level ppf =
+  Fmt_tty.setup_std_outputs ?style_renderer ();
+  Logs.set_level level;
+  Logs.set_reporter (reporter ppf);
+  let quiet = match style_renderer with Some _ -> true | None -> false in
+  quiet, ppf
+
+type error = [ `Store of Store.error | `Sync of Sync.error ]
+
+let store_err err = `Store err
+let sync_err err = `Sync err
+
+let pp_error ppf = function
+  | `Store err -> Fmt.pf ppf "(`Store %a)" Store.pp_error err
+  | `Sync err -> Fmt.pf ppf "(`Sync %a)" Sync.pp_error err
+
+let main (references : (Git.Reference.t * Git.Reference.t) list)
+    (directory : string option) (repository : Uri.t) =
+  let ( >>?= ) = Lwt_result.bind in
+  let ( >>!= ) v f = Lwt_result.map_err f v in
+  let root : Fpath.t =
+    Fpath.v @@ match directory with Some p -> p | None -> Sys.getcwd ()
+  in
+  let resolvers : Conduit.resolvers =
+    Conduit_lwt.empty
+    |> Conduit_lwt.add Conduit_lwt.TCP.protocol
+         (Conduit_lwt.TCP.resolve ~port:80)
+  in
+  let remote_references = List.map fst references in
+  Lwt.return @@ Smart_git.endpoint_of_string (Uri.to_string repository)
+  >>!= store_err
+  >>?= fun endpoint ->
+  Store.v root >>!= store_err >>?= fun store ->
+  Sync.fetch ~resolvers endpoint store (`Some remote_references) >>!= sync_err
```

Comment:
The current behavior with `Git_unix.Sync` is to update a local reference A to what the same reference points out into the server. Currently, it's not _exactly_ what `git fetch` does. In fact, with `git fetch`:
- if you say `A:B`, you will try to get the reference A (which exist into the server) but you will set then locally the reference B
- if you say `A`, this is currently the behavior of `Git_unix.Sync.fetch`
- if you say `A:`, you will save the hash into `remotes/A` (and into `FETCH_HEAD`)
- and `:B` is wrong about fetching if I remember correctly (or, at least, require a more complex context)

It should be good to have the same behavior here - and may be update `Git_unix.Sync` to do properly these operations.